### ISSUE TITLE
Oracle issue - Yuchen

### ISSUE THREAD

**Issue by amva13** [mims-harvard/TDC#238]
**Describe the bug**

Dear TDC Team,
 
I hope this message finds you well. I am writing to report some technical issues I encountered while utilizing the oracle provided by TDC. Below are the details of the problems:
 
Problem 1: I have encountered an error after downloading the Oracle with the name "JNK3". The error message is as follows:
ValueError: node array from the pickle has an incompatible dtype:
- expected: {'names': ['left_child', 'right_child', 'feature', 'threshold', 'impurity', 'n_node_samples', 'weighted_n_node_samples', 'missing_go_to_left'], 'formats': ['<i8', '<i8', '<i8', '<f8', '<f8', '<i8', '<f8', 'u1'], 'offsets': [0, 8, 16, 24, 32, 40, 48, 56], 'itemsize': 64}
- got: [('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]

This problem also occurs with the Oracle named "GSK3". However, the error arises when I attempt to input a list of smiles. Inputting a single smile into GSK3 does not trigger the error.
 
Problem 2: As mentioned above, inputting a single smile into the Oracle "GSK3" does not result in an error. However, I have tried multiple active molecules targeting GSK3beta from ChEMBL, and the output value from the oracle is consistently 0. This suggests there might be an issue with the "GSK3" oracle that requires your attention.
 
I hope you can address these issues promptly. Please let me know if you need any further information or details from my end.
 
Best regards,
 
Yuchen

**Comment by amva13** [comment]
@abearab ^ you can have a look at this

**Comment by miguelgondu** [comment]
I'm having the same issue. I wonder whether something changed in sklearn's random forests and their formatting. That being said, I'm using `sklearn==1.3.0`, which is the version inside this project's `requirements.txt`.

**Comment by miguelgondu** [comment]
The culprit for the discrepancy between lists/individual SMILES is [the try-except block in L656](https://github.com/mims-harvard/TDC/blob/892f4400e38016e69a7ecf687d5b2853b17204c1/tdc/oracles.py#L656) of the implementation of oracles.

In other words, the loading of the oracle is failing silently, and thus the oracle returns the default value.

So we could try to solve two problems:
1. Calling oracles on `smile_str` and `[smile_str]` should have the same behavior.
2. Fixing the loading of the oracles for `GSK3B` and `JNK3`.

I'm happy to volunteer on any of those!

**Comment by amva13** [comment]
Hi @miguelgondu , thanks for the find! For clarity, changing the try-except block would only reveal the real error, not fix it. What version of the package are you using? Could you try 0.4.1 ?

**Comment by amva13** [comment]
Ok. This was to confirm error is not due to recent release changes. I will be personally inspecting this error starting now. One thing I'd try while I'm looking into it. There might be something to your claim about sklearn==1.3.0 causing a breaking change.

I would try building package 0.4.1 in a virtual environment (i.e. conda). 0.4.1 does not specify versions in requirements.txt and this might fix the behavior.

**Comment by amva13** [comment]
This error is indeed because of a mismatch in the formatting between the pickle object and the format expected by scikit learn. This is in part due to a version upgrade in scikit.

See reverse issue here
https://github.com/yzhao062/pyod/issues/519

Evaluating some fixes and will push new version of package asap.

**Comment by amva13** [comment]
Hi @miguelgondu I believe I've solved it. Would you mind sharing some of the input SMILES strings which produced a 0.0 value for these oracles for you?

**Comment by amva13** [comment]
Hi @miguelgondu I just pushed the fix and will be releasing the new package now. Will lyk when you can install

**Comment by miguelgondu** [comment]
Thanks! Looking forward.

**Comment by miguelgondu** [comment]
Just FYI: I'm getting a warning on `Thiothixene_Rediscovery` that is similar in spirit to this issue:

```
InconsistentVersionWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.23.0 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
  https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
```

**Comment by amva13** [comment]
Got it. Thanks for pointing out. The best solution is to pickle these solutions with a more modern scikit. For now the downgrade seems to work, though that particular classifier came from version 0.23.0.. so not great. I'll flag this is a longer term issue to look at.

**Comment by amva13** [comment]
@miguelgondu it's all fixed. you can install 0.4.7 for the working version

example:
https://colab.research.google.com/drive/17mGlLaVkfA2-0sqhbZlQ4cUI0JnFBpRq?usp=sharing

**Comment by miguelgondu** [comment]
Hi @amva13 , thanks for the fix!

Checking with the other oracles in that specific version, something seems to break in `deco hop`. In the first example of the documentation (the same one I provided above) I went from getting `0.5338...` to getting `0.0`. Weird!

The rest of the oracles seem to work as expected, except for the ones in the issue I raised recently (#244).

Thanks again for the hard work.

**Comment by amva13** [comment]
ack'd issue opened

**Comment by kaechb** [comment]
Dear TDC Team,

Firstly, thank you for developing and maintaining such a valuable codebase. I recently encountered an issue with the GSK3B oracle that I believe could benefit from improved error handling.

While utilizing the GSK3B oracle, I observed that it consistently returned a score of 0.0, without any accompanying error messages or warnings. This silent failure led me to initially suspect issues within my model. Upon further investigation, I discovered that the root cause was a compatibility issue with scikit-learn:
```
ValueError: node array from the pickle has an incompatible dtype:
- expected: {'names': ['left_child', 'right_child', 'feature', 'threshold', 'impurity', 'n_node_samples', 'weighted_n_node_samples', 'missing_go_to_left'], ...}
- got     : [('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]
```
This error suggests that the pickled model was created with an older version of scikit-learn, leading to incompatibility with the current version in use.

To enhance the user experience and facilitate easier debugging, I recommend modifying the exception handling within the oracle’s evaluation function. Specifically, catching exceptions without providing feedback can obscure underlying issues. Incorporating a traceback can offer immediate insights into any problems. Here’s a suggested modification:
```
import traceback

try:
    score = self.evaluator_func(*args, **kwargs)
except Exception:
    traceback.print_exc()
    score = self.default_property
```
By implementing this change, users would receive immediate feedback on errors, allowing for quicker diagnosis and resolution.
